target_chains/ethereum/sdk/solidity/Price.sol

```diff
@@ -0,0 +1,459 @@
+     * @param x1 The x coordinate of the first point
+     * @param y1 The y coordinate of the first point, represented as a Price struct
+     * @param x2 The x coordinate of the second point, must be greater than x1
+     * @param y2 The y coordinate of the second point, represented as a Price struct
+     * @param x_query The query x coordinate, at which we wish to impute a y value
+     * @param pre_add_expo The exponent to scale to, before final addition; essentially the final precision you want
+     * @return The price at the query x coordinate
+     */
+    function affine_combination(
+        int64 x1,
+        PythStructs.Price memory y1,
+        int64 x2,
+        PythStructs.Price memory y2,
+        int64 x_query,
+        int32 pre_add_expo
+    ) internal pure returns (PythStructs.Price memory) {
+        if (x2 <= x1) {
+            revert("x2 must be greater than x1");
+        }
+
+        // Get the deltas for the x coordinates
+        // 1. compute A = x_query - x1
+        int64 delta_q1 = x_query - x1;
+        // 2. compute B = x2 - x_query
+        int64 delta_2q = x2 - x_query;
+        // 3. compute C = x2 - x1
+        int64 delta_21 = x2 - x1;
+
+        // Get the relevant fractions of the deltas, with scaling
+        // 4. compute D = A / C
+        PythStructs.Price memory frac_q1 = fraction(delta_q1, delta_21);
+        // 5. compute E = B / C
+        PythStructs.Price memory frac_2q = fraction(delta_2q, delta_21);
+
+        // Calculate products for left and right
+        // 6. compute F = y2 * D
+        PythStructs.Price memory left = mul(y2, frac_q1);
+        // 7. compute G = y1 * E
+        PythStructs.Price memory right = mul(y1, frac_2q);
+
+        // Scaling
+        left = scale_to_exponent(left, pre_add_expo);
+        right = scale_to_exponent(right, pre_add_expo);
+
+        // 8. compute H = F + G
+        return add(left, right);
+    }
+
+    /**
+     * @dev Divide one price by another, propagating uncertainty.
+     * @param self The numerator price
+     * @param other The denominator price
+     * @return The resulting price
+     */
+    function div(
+        PythStructs.Price memory self,
+        PythStructs.Price memory other
+    ) internal pure returns (PythStructs.Price memory) {
+        PythStructs.Price memory base = normalize(self);
+        other = normalize(other);
+
+        if (other.price == 0) {
+            revert DivisionByZero();
+        }
+
+        (uint64 basePrice, int64 baseSign) = toUnsigned(base.price);
+        (uint64 otherPrice, int64 otherSign) = toUnsigned(other.price);
+
+        uint64 midprice = (basePrice * PD_SCALE) / otherPrice;
+        int32 midpriceExpo = base.expo - other.expo + PD_EXPO;
+
+        uint64 otherConfidencePct = (other.conf * PD_SCALE) / otherPrice;
+        uint128 conf = ((uint128(base.conf) * PD_SCALE) / otherPrice) +
```

Comment:
i think solidity handles casting uints up, but make sure? generally need some tests to make sure this all works